command/agent/agent_endpoint_test.go

```diff
@@ -672,60 +550,43 @@ func TestAgent_Leave(t *testing.T) {
 }
 
 func TestAgent_Leave_ACLDeny(t *testing.T) {
-	// Try without a token.
-	func() {
+	t.Run("no token", func(t *testing.T) {
 		dir, srv := makeHTTPServerWithACLs(t)
 		defer os.RemoveAll(dir)
 		defer srv.Shutdown()
 		defer srv.agent.Shutdown()
 
-		req, err := http.NewRequest("PUT", "/v1/agent/leave", nil)
-		if err != nil {
-			t.Fatalf("err: %v", err)
-		}
-
-		_, err = srv.AgentLeave(nil, req)
-		if err == nil || !strings.Contains(err.Error(), permissionDenied) {
+		req, _ := http.NewRequest("PUT", "/v1/agent/leave", nil)
+		if _, err := srv.AgentLeave(nil, req); !isPermissionDenied(err) {
 			t.Fatalf("err: %v", err)
 		}
-	}()
+	})
 
-	// Try the agent master token (resolved on the agent).
-	func() {
+	t.Run("agent master token", func(t *testing.T) {
 		dir, srv := makeHTTPServerWithACLs(t)
 		defer os.RemoveAll(dir)
 		defer srv.Shutdown()
 		defer srv.agent.Shutdown()
 
-		req, err := http.NewRequest("PUT", "/v1/agent/leave?token=towel", nil)
-		if err != nil {
-			t.Fatalf("err: %v", err)
-		}
-
-		_, err = srv.AgentLeave(nil, req)
-		if err != nil {
+		req, _ := http.NewRequest("PUT", "/v1/agent/leave?token=towel", nil)
+		if _, err := srv.AgentLeave(nil, req); err != nil {
 			t.Fatalf("err: %v", err)
 		}
-	}()
+	})
 
-	// Try with a read only token (resolved on the servers).
-	func() {
+	// this test fails when all subtests share the same server.
```

Comment:
When I refactor the code like this:

```go
func TestAgent_Leave_ACLDeny(t *testing.T) {
	dir, srv := makeHTTPServerWithACLs(t)
	defer os.RemoveAll(dir)
	defer srv.Shutdown()
	defer srv.agent.Shutdown()

	t.Run("no token", func(t *testing.T) {
		req, _ := http.NewRequest("PUT", "/v1/agent/leave", nil)
		if _, err := srv.AgentLeave(nil, req); !isPermissionDenied(err) {
			t.Fatalf("err: %v", err)
		}
	})

	t.Run("agent master token", func(t *testing.T) {
		req, _ := http.NewRequest("PUT", "/v1/agent/leave?token=towel", nil)
		if _, err := srv.AgentLeave(nil, req); err != nil {
			t.Fatalf("err: %v", err)
		}
	})

	// this test fails when all subtests share the same server.
	t.Run("read-only token", func(t *testing.T) {
		ro := makeReadOnlyAgentACL(t, srv)
		req, _ := http.NewRequest("PUT", fmt.Sprintf("/v1/agent/leave?token=%s", ro), nil)
		if _, err := srv.AgentLeave(nil, req); !isPermissionDenied(err) {
			t.Fatalf("err: %v", err)
		}
	})
}
```

I get this:

```shell
$ go test -v -run TestAgent_Leave_ACLDeny
=== RUN   TestAgent_Leave_ACLDeny
2017/05/10 08:58:06 [INFO] raft: Initial configuration (index=1): [{Suffrage:Voter ID:127.0.0.1:20011 Address:127.0.0.1:20011}]
2017/05/10 08:58:06 [INFO] raft: Node at 127.0.0.1:20011 [Follower] entering Follower state (Leader: "")
2017/05/10 08:58:06 [INFO] serf: EventMemberJoin: Node 10006 127.0.0.1
2017/05/10 08:58:06 [INFO] consul: Adding LAN server Node 10006 (Addr: tcp/127.0.0.1:20011) (DC: dc1)
2017/05/10 08:58:06 [INFO] serf: EventMemberJoin: Node 10006.dc1 127.0.0.1
2017/05/10 08:58:06 [INFO] consul: Handled member-join event for server "Node 10006.dc1" in area "wan"
2017/05/10 08:58:06 [WARN] raft: Heartbeat timeout from "" reached, starting election
2017/05/10 08:58:06 [INFO] raft: Node at 127.0.0.1:20011 [Candidate] entering Candidate state in term 2
2017/05/10 08:58:06 [DEBUG] raft: Votes needed: 1
2017/05/10 08:58:06 [DEBUG] raft: Vote granted from 127.0.0.1:20011 in term 2. Tally: 1
2017/05/10 08:58:06 [INFO] raft: Election won. Tally: 1
2017/05/10 08:58:06 [INFO] raft: Node at 127.0.0.1:20011 [Leader] entering Leader state
2017/05/10 08:58:06 [INFO] consul: cluster leadership acquired
2017/05/10 08:58:06 [INFO] consul: New leader elected: Node 10006
2017/05/10 08:58:06 [DEBUG] consul: reset tombstone GC to index 3
2017/05/10 08:58:06 [INFO] consul: member 'Node 10006' joined, marking health alive
2017/05/10 08:58:06 [DEBUG] consul: dropping node "Node 10006" from result due to ACLs
=== RUN   TestAgent_Leave_ACLDeny/no_token
=== RUN   TestAgent_Leave_ACLDeny/agent_master_token
2017/05/10 08:58:06 [INFO] consul: server starting leave
2017/05/10 08:58:06 [INFO] serf: EventMemberLeave: Node 10006.dc1 127.0.0.1
2017/05/10 08:58:06 [INFO] serf: EventMemberLeave: Node 10006 127.0.0.1
2017/05/10 08:58:06 [INFO] agent: requesting shutdown
2017/05/10 08:58:06 [INFO] consul: shutting down server
2017/05/10 08:58:06 [INFO] consul: Handled member-leave event for server "Node 10006.dc1" in area "wan"
2017/05/10 08:58:06 [INFO] manager: shutting down
2017/05/10 08:58:06 [INFO] agent: shutdown complete
=== RUN   TestAgent_Leave_ACLDeny/read-only_token
2017/05/10 08:58:06 [DEBUG] http: Shutting down http server (127.0.0.1:20008)
--- FAIL: TestAgent_Leave_ACLDeny (0.33s)
    --- PASS: TestAgent_Leave_ACLDeny/no_token (0.00s)
    --- PASS: TestAgent_Leave_ACLDeny/agent_master_token (0.00s)
    --- FAIL: TestAgent_Leave_ACLDeny/read-only_token (0.00s)
    	agent_endpoint_test.go:36: err: No cluster leader
FAIL
exit status 1
FAIL	github.com/hashicorp/consul/command/agent	0.388s
```

When I re-order the tests then it works:

```go
func TestAgent_Leave_ACLDeny(t *testing.T) {
	dir, srv := makeHTTPServerWithACLs(t)
	defer os.RemoveAll(dir)
	defer srv.Shutdown()
	defer srv.agent.Shutdown()

	t.Run("no token", func(t *testing.T) {
		req, _ := http.NewRequest("PUT", "/v1/agent/leave", nil)
		if _, err := srv.AgentLeave(nil, req); !isPermissionDenied(err) {
			t.Fatalf("err: %v", err)
		}
	})

	t.Run("read-only token", func(t *testing.T) {
		ro := makeReadOnlyAgentACL(t, srv)
		req, _ := http.NewRequest("PUT", fmt.Sprintf("/v1/agent/leave?token=%s", ro), nil)
		if _, err := srv.AgentLeave(nil, req); !isPermissionDenied(err) {
			t.Fatalf("err: %v", err)
		}
	})

	// this sub-test will change the state so that there is no leader.
	// it must therefore be the last one in this list.
	t.Run("agent master token", func(t *testing.T) {
		req, _ := http.NewRequest("PUT", "/v1/agent/leave?token=towel", nil)
		if _, err := srv.AgentLeave(nil, req); err != nil {
			t.Fatalf("err: %v", err)
		}
	})
}
```

```shell
--- PASS: TestAgent_Leave_ACLDeny (0.22s)
    --- PASS: TestAgent_Leave_ACLDeny/no_token (0.00s)
    --- PASS: TestAgent_Leave_ACLDeny/read-only_token (0.00s)
    --- PASS: TestAgent_Leave_ACLDeny/agent_master_token (0.00s)
PASS
```

This feels like they are three separate tests and we can fit them into a single test function. Either we start a separate server for each sub test or we depend on the order of the tests since the last call succeeds and changes the state.